### DISCUSSION TITLE
WITH statement to support variables.

### DISCUSSION THREAD

**Discussion by rustyconover** [duckdb/duckdb#14155] [Ideas]
I'd love to explore adding support for [variables](https://duckdb.org/docs/sql/statements/set_variable.html) within the `WITH` statement, as I think it could be a valuable feature.

Here’s an example of how it might look:

```sql
WITH VARIABLE buckets (SELECT ['a', 'b', 'c']) 
SELECT getvariable('buckets')
```

In terms of scoping, the idea is that the variable would only be accessible within the context of the query. Once the query completes, the variable would no longer be available.

I’d be happy to hear any thoughts or feedback on this!

**Comment by ajzo90** [comment]
This would "solve" an issue I have where I "need" to choose between convenience and performance. With this I would get both. https://discord.com/channels/909674491309850675/921073327009853451/1281904020625162304

Taking it a bit further, it would also be convenient to be able to use macros (and lambdas?) similarly.
